### PR TITLE
Make changes to Style numFmtId parser for excel

### PR DESCRIPTION
**Pull request by semibroiled**
### SUMMARY
Implements necessary changes for #151 

*Reviewers:* @numere-org/maintainers

### IMPLEMENTATION
* Implementation: in the file.hpp file, I've imported in std::set and std::find_if by adding std::<set> and std::<algorithm>
in the file.cpp file, I've made 3 changes primarily
- 1. in the XLSXSpreadSheet::readFile() method, I've added in a segment that parses through xl/style.xml to get the corresponding numFmtId in its nodes if its a date time format. Jump to this section by searching comment // Parse through style.xml to get numFmtId
- 2. I've added in a preceding function thats checks the format code Attribute to determine if it's a date time format. I do this by checking if the character m, h, d, j, s and their capitalised counterpart characters are in the formatCode attribute. isDateTimeFormat for more information. I'm not sure if this function is best added in the .hpp file or as is
- 3. In the same readFile() Function, instead of checking hardcoded Ids, if check to see if the style id is in my vDateTimeFormat set. If it is something should be done in convertExcelTimetoEpoch function. I wasn't sure what the todos here would be. Should we always strive to get seconds since xx-xx-xx and convert to date as mm-dd-jjjj syntax or each excel formatCode should be represented? 
- 
* Implementation test: In the SW_test suite, there is an xlsx file INVENSITY Public (C0) that contains all implementations of date time format found in Date, Time and User Defined. I have not made sensible asserts as they do not pass, or it is a problem with the code. All other tests pass, so I haven't broken anything else. The code should compile without additional errors

### DOCUMENTATION
* [x] ChangesLog updated
* [x] Code changes commented
* **Documentation articles:**
    * [ ] corresponding documentation articles updated
    * [ ] new documentation articles created
    * [x] not needed
* **Language files:**
    * [ ] corresponding language files updated
    * [x] not needed

----------------

### TESTS BY REVIEWERS
* [x] Added to the automatic SW tests
* [ ] Tested manually

### CODE REVIEW

File: kernel/core/io/file.cpp [hunk 1]
```diff
@@ -3899,6 +3899,33 @@ namespace NumeRe
...
+    /// code to be checked.
+    ///
+    /// \return true if the formatString is a date-time format,
+    /// false otherwise.
+    ///
+    /////////////////////////////////////////////////
+    bool isDateTimeFormat(const std::string& formatString)
+    {
+        // List of characters to search for
+        const char charsToFind[] = {'m', 'd', 's', 'h', 'M', 'D', 'S', 'H'};
```
Comment (semibroiled) [inline kernel/core/io/file.cpp:3920]: For some reasons I cant convert to Draft mode. I need to add j and J here. Maybe adding English variants like y or Y also makes sense for this case.

Is this function okay as is? or best written in .hpp?

File: kernel/core/io/file.cpp [hunk 2]
```diff
@@ -4117,6 +4144,31 @@ namespace NumeRe
         _styles.Parse(sStylesContent.c_str());
         tinyxml2::XMLElement* cellXfs = _styles.FirstChildElement()->FirstChildElement("cellXfs");
 
+        // Parse through styles.xml to get numfmts
+        // Iterate through nodes
+        tinyxml2::XMLElement *numFmts = _styles.FirstChildElement()->FirstChildElement("numfmts");
+
+        // Empty set to store numfmts ID
+        std::set<int> vDateTimeIds;
```
Comment (semibroiled) [inline kernel/core/io/file.cpp:4152]: Set to get unique numFmtIds

File: kernel/core/io/file.cpp [hunk 3]
```diff
@@ -4117,6 +4144,31 @@ namespace NumeRe
...
+            numtFmts = numFmts->FirstChildElement();
+
+            while (numFmts)
+            {
+                int id = numFmts->IntAttribute("numFmtId");
+                std::string formatString = numFmts->Attribute("formatCode");
+
+                // Detect datetime, if yes, push to empty set
+                if (isDateTimeFormat(formatString){
+                    vDateTimeIds.insert(id); // Insert into set
```
Comment (semibroiled) [inline kernel/core/io/file.cpp:4165]: Push to set if formatCode attribute is of date time type. All formatCode date times have mmm-yyy of some kind

File: kernel/core/io/file.cpp [hunk 4]
```diff
@@ -4241,10 +4293,14 @@ namespace NumeRe
                             {
                                 styleId = style->IntAttribute("numFmtId");
 
-                                // Those are time formats: 14-22, 45-47
-                                if ((styleId >= 14 && styleId <= 22) || (styleId >= 45 && styleId <= 47))
+                                // Check if styleId is in vDateTimeIds set
+                                if (vDateTimeIds.find(styleId) != vDateTimeIds.end())
```
Comment (semibroiled) [inline kernel/core/io/file.cpp:4297]: Instead of hardcoded checking, see if my set has any of styleID

File: kernel/core/io/file.cpp [hunk 5]
```diff
@@ -4241,10 +4293,14 @@ namespace NumeRe
                             {
                                 styleId = style->IntAttribute("numFmtId");
 
-                                // Those are time formats: 14-22, 45-47
-                                if ((styleId >= 14 && styleId <= 22) || (styleId >= 45 && styleId <= 47))
+                                // Check if styleId is in vDateTimeIds set
+                                if (vDateTimeIds.find(styleId) != vDateTimeIds.end())
+                                {
+                                    // What change can i make to convertExcelTimeEpoch function?
                                     sValue = convertExcelTimeToEpoch(sValue, (styleId >= 18 && styleId <= 21)
```
Comment (semibroiled) [inline kernel/core/io/file.cpp:4300]: I am unsure about convertExcelTimeToEpoch. We haven't talked about this in detail yet, or much at all. It should be integral to date time functionality

File: kernel/core/io/file.hpp [hunk 6]
```diff
@@ -24,6 +24,8 @@
 #include <cmath>
 #include <vector>
 #include <utility>
+#include <set>
```
Comment (semibroiled) [inline kernel/core/io/file.hpp:27]: in hpp importing set and algorithm for find_if. Looks to be fastest way to do either of its kind

**Comment by semibroiled** [comment]
I'm unable to enter draft mode to do a commit. I want to make changes to changelog and add more characters to isDateTime... Function.

I haven't tested with SW_Test suite yet. I haven't broken anything else. But I couldn't pass or make sensible integration tests. I will follow up on this in the meeting. I suspect changes to changeExcelTimetoEpoch needs to be made

File: kernel/core/io/file.cpp (hunk 4 above)
Comment (numeredev) [reply kernel/core/io/file.cpp:4297]: Your set probably won't contain the standard IDs as written in the comment:
`// Those are time formats: 14-22, 45-47`
I'd recommend to use them together with your solution like so

```C++
if ((styleId >= 14 && styleId <= 22) 
    || (styleId >= 45 && styleId <= 47) 
    || vDateTimeIds.find(styleId) != vDateTimeIds.end())
```

File: kernel/core/io/file.cpp (hunk 5 above)
Comment (numeredev) [reply kernel/core/io/file.cpp:4300]: It is. It converts from 1.1.1900 to the UNIX epoch (i.e. 1.1.1970).

File: kernel/core/io/file.hpp (hunk 6 above)
Comment (numeredev) [reply kernel/core/io/file.hpp:27]: There's probably no need to have it in the hpp file, as it is only referenced in file.cpp. Move it to this file please (keep foreign imports* as clean as possible)

* Foreign imports: Imports that you receive via an imported file.

File: kernel/core/io/file.cpp [hunk 7]
```diff
@@ -4241,10 +4293,14 @@ namespace NumeRe
...
 
-                                // Those are time formats: 14-22, 45-47
-                                if ((styleId >= 14 && styleId <= 22) || (styleId >= 45 && styleId <= 47))
+                                // Check if styleId is in vDateTimeIds set
+                                if (vDateTimeIds.find(styleId) != vDateTimeIds.end())
+                                {
+                                    // What change can i make to convertExcelTimeEpoch function?
                                     sValue = convertExcelTimeToEpoch(sValue, (styleId >= 18 && styleId <= 21)
+                                
                                                                               || (styleId >= 45 && styleId >= 47));
```
Comment (numeredev) [inline kernel/core/io/file.cpp:4302]:
```suggestion
                                                                              || (styleId >= 45 && styleId <= 47));
```

File: kernel/core/io/file.cpp (hunk 1 above)
Comment (numeredev) [inline kernel/core/io/file.cpp]:
```suggestion
    static bool isDateTimeFormat(const std::string& formatString)
    {
        // List of characters to search for
        static const char charsToFind[] = {'m', 'd', 's', 'h', 'M', 'D', 'S', 'H'};
```

File: kernel/core/io/file.cpp [hunk 8]
```diff
@@ -4241,10 +4300,15 @@ namespace NumeRe
...
-                                    sValue = convertExcelTimeToEpoch(sValue, (styleId >= 18 && styleId <= 21)
-                                                                              || (styleId >= 45 && styleId >= 47));
+                                // Check if styleId is in vDateTimeIds or vStandardDateTimeIds set
+                                if ((vDateTimeIds.find(styleId) != vDateTimeIds.end()) || 
+                                (vStandardDateTimeIds.find(styleId) != vStandardDateTimeIds.end()))
+                                {
+
+                                    //? Is it standard time or date time? What about mixed values with date and time?
+                                    sValue = convertExcelTimeToEpoch(sValue, 
+                                    (vStandardDateTimeIds.find(styleId) != vStandardDateTimeIds.end()));
```
Comment (numeredev) [inline kernel/core/io/file.cpp:4310]: Here, a possible problem will arise: the second parameter is a boolean flag for "value is only time value and does not contain date information". This is not true for the standard IDs 14-17 (as visible in the old implementation).

I recommend to have separate `std::set`s for date values and time values. That's kind of a duplicate, but yeah, it's Excel.

File: kernel/core/io/file.cpp [hunk 9]
```diff
@@ -4117,6 +4147,35 @@ namespace NumeRe
         _styles.Parse(sStylesContent.c_str());
         tinyxml2::XMLElement* cellXfs = _styles.FirstChildElement()->FirstChildElement("cellXfs");
 
+        // Initialize Set with Standard Excel Datetime IDs: 18..21, 45..47
+        std::set<int> vStandardDateTimeIds = {14, 15, 16, 17, 18, 19, 20, 21, 22, 45, 46, 47};
```
Comment (numeredev) [inline kernel/core/io/file.cpp:4151]: As said before, I recommend to have separate sets for time and date values (14-17 are date, the others are time values)

File: kernel/core/io/file.cpp [hunk 10]
```diff
@@ -3899,6 +3902,33 @@ namespace NumeRe
...
+    /// code to be checked.
+    ///
+    /// \return true if the formatString is a date-time format,
+    /// false otherwise.
+    ///
+    /////////////////////////////////////////////////
+    bool isDateTimeFormat(const std::string& formatString)
+    {
+        // List of characters to search for
+        const char charsToFind[] = {'m', 'd', 's', 'h','j', 'M', 'D', 'S', 'H', 'J'};
```
Comment (numeredev) [inline kernel/core/io/file.cpp:3923]: It is probably reasonable to have two functions: `isDateFormat(...)` and `isTimeFormat(...)`.

**Review by numeredev — CHANGES_REQUESTED** [review]
With exception of the issue regarding the separation of date and time values, it looks already quite good!

File: kernel/core/io/file.cpp [hunk 11]
```diff
@@ -4293,10 +4400,18 @@ namespace NumeRe
...
                                 styleId = style->IntAttribute("numFmtId");
 
-                                // Those are time formats: 14-22, 45-47
-                                if ((styleId >= 14 && styleId <= 22) || (styleId >= 45 && styleId <= 47))
-                                    sValue = convertExcelTimeToEpoch(sValue, (styleId >= 18 && styleId <= 21)
-                                                                              || (styleId >= 45 && styleId >= 47));
+                                // Check if styleId is in vTimeIds or vStandardTimeIds sets
+                                isTime = (vTimeIds.find(styleId) != vTimeIds.end()) ||
+                                    (vStandardTimeIds.find(styleId) != vStandardTimeIds.end()); 
+
```
Comment (numeredev) [inline kernel/core/io/file.cpp:4406]:
```suggestion
                                bool isTime = vTimeIds.find(styleId) != vTimeIds.end()
                                    || vStandardTimeIds.find(styleId) != vStandardTimeIds.end(); 
                                bool isDate = vDateIds.find(styleId) != vDateIds.end()
                                    || vStandardDateIds.find(styleId) != vStandardDateds.end(); 
```

File: kernel/core/io/file.cpp [hunk 12]
```diff
@@ -4293,10 +4400,18 @@ namespace NumeRe
...
+                                
+                                if (isTime)
```
Comment (numeredev) [inline kernel/core/io/file.cpp:4408]:
```suggestion
                                if (isDate || isTime)
```

File: kernel/core/io/file.cpp [hunk 13]
```diff
@@ -3951,6 +3954,68 @@ namespace NumeRe
...
+    /// code to be checked.
+    ///
+    /// \return true if the formatString is a date format,
+    /// false otherwise.
+    ///
+    /////////////////////////////////////////////////
+    bool isDateFormat(const std::string &formatString)
+    {
+        // List of characters to search for
+        const char charsToFind[] = {'m', 'M', 'd', 'D', 'j', 'J', 'y', 'Y'};
```
Comment (numeredev) [inline kernel/core/io/file.cpp:3978]:
```suggestion
    static bool isDateFormat(const std::string &formatString)
    {
        // List of characters to search for
        static const char charsToFind[] = {'m', 'M', 'd', 'D', 'j', 'J', 'y', 'Y'};
```

File: kernel/core/io/file.cpp [hunk 14]
```diff
@@ -3951,6 +3954,68 @@ namespace NumeRe
...
+    /// code to be checked.
+    ///
+    /// \return true if the formatString is a time format,
+    /// false otherwise.
+    ///
+    /////////////////////////////////////////////////
+    bool isTimeFormat(const std::string &formatString)
+    {
+        // List of characters to search for
+        const char charsToFind[] = {'s', 'S', 'h', 'H'};
```
Comment (numeredev) [inline kernel/core/io/file.cpp:4010]:
```suggestion
    static bool isTimeFormat(const std::string &formatString)
    {
        // List of characters to search for
        static const char charsToFind[] = {'s', 'S', 'h', 'H'};
```

File: kernel/core/io/file.cpp [hunk 15]
```diff
@@ -4169,6 +4234,48 @@ namespace NumeRe
...
+                    vDateIds.insert(id); // Insert into set
+                };
+
+                numFmts = numFmts->NextSiblingElement();
+            }
+        }
+
+        // Empty boolean value to store Time format inspection
+        bool isTime;
+
```
Comment (numeredev) [inline kernel/core/io/file.cpp:4278]:
```suggestion
```

**Review by numeredev — CHANGES_REQUESTED** [review]
Thanks for the changes, only minor things remain and the central question, whether the time and date chars are correct (I'd assume that only english notation is used within the styles.xml).

**Review by numeredev — APPROVED** [review]

**Pull request merged**